File: Appointment.java
// Author Name: Owen Lovell

// Date: 10/06/2024

// Course ID: CS-320-13376-M01

// Description: This is the Appointment class that creates an appointment object that holds
// a appointmentID, a date, and description.

package module5;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.concurrent.atomic.AtomicLong;

public class Appointment {
	
	private final String appointmentID;
	private Date appointmentDate;
	private String description;
	private static AtomicLong idGenerator = new AtomicLong();
	
	public Appointment(String dateOfAppointment, String description) {
		this.appointmentID = String.valueOf(idGenerator.getAndIncrement());
		 
		// DATE - day of today set to check if appointment date is in the past
		// fist if appointment is null then todays date will be auto assigned
		SimpleDateFormat formatter = new SimpleDateFormat("MM/dd/YYYY"); 
		Date todaysDate = new Date(10/6/2024);	// FIXME: date is hardcoded
		
		if(dateOfAppointment == null) {
			this.appointmentDate = todaysDate;
		} else {
			try {
				Date temp = formatter.parse(dateOfAppointment);
				if(temp.before(todaysDate) == true) { 
					System.out.println("Date in the past!");
				} else {
					this.appointmentDate = temp;
				}
			} catch (ParseException e) {
				System.out.println("String DATE unable to be converted into date.");
				e.printStackTrace();
			}
		}
				
		// DESCRIPTION
		if(description == null || description.isBlank()){
			this.description = "NULL";
		} else if(description.length() > 50) {
			//System.out.println("Description too long, only first 50 characters kept.");
			this.description = description.substring(0,50);
		} else {
			this.description = description;
		}
	}
	
	// Getters
	public String getAppointmentID() {
		return appointmentID;
	}
	
	public Date getAppointmentDate() {
		return appointmentDate;
	}
	
	public String getDescription() {
		return description;
	}
	
	// Setters
	
	public void setAppointmentDate(String newDate) {
		SimpleDateFormat formatter = new SimpleDateFormat("MM/dd/YYYY");
		Date todaysDate = new Date(10/6/2024);
		try {
			Date temp = formatter.parse(newDate);
			if(temp.before(todaysDate) == true) { 
				System.out.println("Date in the past!");
			} else {
				this.appointmentDate = temp;
			}
		} catch (ParseException e) {
			System.out.println("String DATE unable to be converted into date.");
			e.printStackTrace();
		}
	}
	
	public void setDescription(String newDescription) {
		if(newDescription.length() > 50) {
			//System.out.println("Description too long, only first 50 characters kept.");
			description = newDescription.substring(0,50);
		} else if(newDescription == null || newDescription.isBlank()){
			description = "NULL";
		} else {
			description = newDescription;
		}
	}

}

File: Projects/Enhancement-2/Enhanced/appointment.py
# Author: Owen Lovell
# Date Updated: 12/8/2025
# Course ID: CS-499-17242-M01

"""
Description: The Appointment is used to create an appointment object, the object holds appointmentID,
a date, and description.
"""
import itertools


class Appointment:

    id_iter = itertools.count()

    # Constructor for appointment date, and description
    def __init__(self, appointment_date, description):

        self.appointment_date = appointment_date
        self.description = description

        self.appointment_id = next(self.id_iter)

    # Getters
    def get_appointment_id(self):
        return self.appointment_id

    def get_date(self):
        return self.appointment_date

    def get_description(self):
        return self.description

    # setDate
    def set_date(self, new_appointment_date):
        self.appointment_date = new_appointment_date

    # setDescription
    def set_description(self, new_description):
        self.description = new_description

    # to string
    def __str__(self):
        return "Appointment ID: {}, Date: {}, Description: {}"\
            .format(self.appointment_id, self.appointment_date, self.description)
